File: LP2/Greedy/JSP.py
# JOB SCHEDULING PROBLEM USING GREEDY METHOD

# Each job has:
# (Job ID, Deadline, Profit)
jobs = [
    ('J1', 2, 100),
    ('J2', 1, 50),
    ('J3', 2, 10),
    ('J4', 1, 20),
    ('J5', 3, 30)
]

# JOB SCHEDULING FUNCTION
def job_scheduling(jobs):

    # Sort jobs in descending order of profit
    jobs.sort(key=lambda x: x[2], reverse=True)

    # Find maximum deadline
    max_deadline = max(job[1] for job in jobs)

    # Create slots for scheduling jobs
    slots = [None] * max_deadline

    # Store total profit
    total_profit = 0

    print("Selected Jobs:\n")

    # Traverse all jobs
    for job_id, deadline, profit in jobs:

        # Check slots from deadline-1 to 0
        for i in range(deadline - 1, -1, -1):

            # If slot is empty
            if slots[i] is None:

                # Assign job to slot
                slots[i] = job_id

                # Add profit
                total_profit += profit

                break

    # Print scheduled jobs
    for i in range(max_deadline):

        if slots[i]:

            print(
                "Time Slot",
                i + 1,
                ":",
                slots[i]
            )

    # Print total profit
    print("\nTotal Profit:", total_profit)


# DRIVER CODE
job_scheduling(jobs)

# time complexity: O(n log n) due to sorting
# space complexity: O(d) where d is the maximum deadline

File: cpp/JSP.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

struct Job {
    char id;
    int deadline;
    int profit;
};

// Compare by profit descending
bool compare(Job a, Job b) {
    return a.profit > b.profit;
}

int main() {

    vector<Job> jobs = {

        {'A', 2, 100},
        {'B', 1, 19},
        {'C', 2, 27},
        {'D', 1, 25},
        {'E', 3, 15}
    };

    // Sort by maximum profit
    sort(jobs.begin(), jobs.end(), compare);

    int n = jobs.size();

    vector<int> slot(n, -1);

    int totalProfit = 0;

    cout << "Selected Jobs:\n";

    for (auto job : jobs) {

        // Find free slot before deadline
        for (int j = job.deadline - 1; j >= 0; j--) {

            if (slot[j] == -1) {

                slot[j] = job.id;

                totalProfit += job.profit;

                cout << job.id << " ";

                break;
            }
        }
    }

    cout << "\n\nMaximum Profit = "
         << totalProfit << endl;

    return 0;
}